### PR TITLE
MetadataManager init

### PR DESCRIPTION
**Pull request by AhmedSoliman**
MetadataManager init

This introduces `MetadataManager`, a node-level service that provides uniform access to different metadata kinds. This PR focuses on the nodes configuration only.

`Metadata` is a cheaply cloneable handle that components can use to access metadata. `MetadataWriter` is only shared with components that can propose updates to the cached metadata.

Additionally, this changes the node startup procedure to be one step closer to the original design. Worker attachment is only concerned with partition assignment and node IDs are directly fetched from NodesConfiguration.

In this, nodes configuration will be statically created on startup if the bootstrap_cluster option is set (set by default at the moment). The server will fail to start if this is unset until we implement fetching nodes configuration from metadata store.

The admin_address is not used anymore on startup but will be replaced with metadata address later for remote nodes config fetching.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1202).
* #1203
* __->__ #1202

### CODE REVIEW

File: crates/node/src/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        loop {
+            tokio::select! {
+            biased;
+            _ = cancellation_watcher() => {
+                info!("Metadata manager stopped");
+                break;
+            }
+            Some(cmd) = self.inbound.recv() => {
+                self.handle_command(cmd)
+            }
```
Comment (tillrohrmann) [inline crates/node/src/metadata.rs:201]: Indentation looks a bit off. Might be GitHub.

File: crates/network/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+use tonic::transport::{Channel, Endpoint};
+use tower::service_fn;
+
+pub fn create_grpc_channel_from_network_address(
+    address: AdvertisedAddress,
+) -> Result<Channel, http::Error> {
+    let channel = match address {
+        AdvertisedAddress::Uds(uds_path) => {
+            // dummy endpoint required to specify an uds connector, it is not used anywhere
+            Endpoint::try_from("/")
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/network/src/utils.rs]: Will fix. I think I just moved this code over, but it's a good catch.

File: crates/node/src/metadata.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/node/src/metadata.rs]: Macros and rustfmt don't like each other :)

Will indent.

File: crates/node/src/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    pub fn metadata(&self) -> Metadata {
+        Metadata::new(self.inner.clone(), self.self_sender.clone())
+    }
+
+    pub fn writer(&self) -> MetadataWriter {
+        MetadataWriter::new(self.self_sender.clone())
+    }
+
+    /// Start and wait for shutdown signal.
+    pub async fn run(mut self /*, network_sender: NetworkSender*/) -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/node/src/metadata.rs:189]: In a possible future, yes.

**Pull request merged**